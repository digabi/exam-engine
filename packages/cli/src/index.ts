#!/usr/bin/env node

import { accessSync } from 'fs'
import ora = require('ora')
import yargs from 'yargs'
import { resolveExam, resolveFile } from './utils'

/* eslint-disable @typescript-eslint/no-misused-promises */
yargs
  .usage('Usage: $0 <command> [options]')
  .command(
    'new <directory>',
    'Create a new exam',
    (argv) => {
      argv
        .positional('directory', {
          description: 'The directory for the new exam',
          type: 'string',
        })
        .demandOption('directory')
    },
    runCommand('./commands/new')
  )
  .command(
    ['preview [exam] [options]', 'start'],
    'Preview an exam',
    (argv) => {
      addExamArgs(argv)
      argv.option('port', {
        alias: 'p',
        description: 'The port to use in the preview HTTP server',
        type: 'number',
        default: 0,
      })
    },
    runCommand('./commands/preview')
  )
  .command(
    'create-transfer-zip [exam] [options]',
    'Create a transfer zip that can be imported to Oma Abitti',
    addExamArgs,
    runCommand('./commands/create-transfer-zip')
  )
  .command(
    'create-offline [exam] [options]',
    'Create a standalone offline version of the exam.',
    (argv) => {
      addExamAndOutdirArgs(argv)
      argv.option('media', {
        description:
          'Create a media version of the exam. This will encode video files as x264 and audio files as mp3 and not remove hidden references from the exam.',
        type: 'boolean',
      })
    },
    runCommand('./commands/create-offline')
  )
  .command(
    'create-mex [exam] [options]',
    'Package the exam to a .mex file that can be imported by Abitti',
    (argv) => {
      addExamAndOutdirArgs(argv)
      argv
        .option('private-key', {
          alias: 'k',
          description: 'The private key (in PEM format) used for signing the exam',
          coerce: resolveFile,
        })
        .option('passphrase', {
          alias: 'p',
          description: 'The secret passphrase that decrypts the exam',
        })
        .option('nsa-scripts', {
          alias: 'n',
          description: 'Monitoring scripts as a .zip file',
          coerce: resolveFile,
        })
        .option('security-codes', {
          alias: 's',
          description: 'The security codes (in JSON format)',
          coerce: resolveFile,
        })
        .demandOption(['private-key', 'passphrase', 'nsa-scripts'])
    },
    runCommand('./commands/create-mex')
  )
  .check((argv: { exam?: string }) => {
    if (argv.exam) {
      accessSync(argv.exam)
    }
    return true
  })
  .demandCommand()
  .strict()
  .wrap(yargs.terminalWidth()).argv

/* eslint-enable @typescript-eslint/no-misused-promises */

function addExamArgs(yargv: yargs.Argv) {
  yargv.positional('exam', {
    description: 'The path to an exam XML file or the directory containing it.',
    default: process.cwd(),
    coerce: resolveExam,
  })
}

function addExamAndOutdirArgs(argv: yargs.Argv) {
  addExamArgs(argv)
  argv.option('outdir', {
    alias: 'o',
    description: 'The output directory',
    type: 'string',
    coerce: resolveFile,
  })
}

function runCommand<T>(moduleName: string) {
  return async (args: T) => {
    const spinner = ora().start()
    try {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
      const command = require(moduleName).default
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-call
      const result = await command({ ...args, spinner })
      if (result) {
        spinner.succeed(result)
      }
    } catch (err) {
      spinner.fail((err as Error).stack)
      process.exit(1)
    } finally {
      spinner.stop()
    }
  }
}
