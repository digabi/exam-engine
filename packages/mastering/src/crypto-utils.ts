import crypto from 'crypto'
import { Readable } from 'stream'

const symmetricAlgo = 'aes-256-ctr'
const signAlgo = 'RSA-SHA256'
const defaultDigest = 'SHA1'

export interface KeyAndIv {
  key: Buffer
  iv: Buffer
}

export function createAES256EncryptStreamWithIv({ key, iv }: KeyAndIv): crypto.Cipher {
  return crypto.createCipheriv(symmetricAlgo, asBuffer(key), asBuffer(iv))
}

export function signWithSHA256AndRSA(input: NodeJS.ReadableStream, privateKeyPem: string): Readable {
  const signer = crypto.createSign(signAlgo)
  const signerStream = input.pipe(signer)
  const output = new Readable()

  signerStream.on('finish', () => {
    output.push(signer.sign(privateKeyPem, 'base64'))
    output.push(null)
  })
  signerStream.on('error', (error) => output.destroy(error))

  return output
}

export function verifyWithSHA256AndRSA(signedDataBuffer: Buffer, publicKey: string, signature: string): boolean {
  const verifier = crypto.createVerify(signAlgo)
  verifier.update(signedDataBuffer)
  return verifier.verify(publicKey, signature, 'base64')
}

export function deriveAES256KeyAndIv(password: string): KeyAndIv {
  const trimmedPassword = password.replace(/\s/g, '')
  const derivedData = crypto.pbkdf2Sync(trimmedPassword, trimmedPassword, 2000, 32 + 16, defaultDigest) // Use password as salt as well, 32 bytes for key, 16 for iv
  const key = derivedData.slice(0, 32)
  const iv = derivedData.slice(32, 48)
  return { key, iv }
}

function asBuffer(val: string | Buffer): Buffer {
  return Buffer.isBuffer(val) ? val : Buffer.from(val, 'base64')
}
