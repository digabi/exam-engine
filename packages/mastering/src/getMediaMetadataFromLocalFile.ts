import ffprobe from 'ffprobe'
import ffprobeStatic from '@ffprobe-installer/ffprobe'

interface ImageMetadata {
  width: number
  height: number
}
interface AudioMetadata {
  duration: number
}

/** Loads media metadata from files on the local file system. */
export const getMediaMetadataFromLocalFile =
  (resolveAttachment: (filename: string) => string) =>
  async (src: string, type: 'video' | 'audio' | 'image'): Promise<ImageMetadata | AudioMetadata> => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-argument
    const parseWebmDuration = (s: any) => parseFloat(s.tags.DURATION.replace(/:/g, ''))
    const path = resolveAttachment(src)

    // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment
    const result = await ffprobe(path, { path: ffprobeStatic.path })
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
    const stream = result.streams[0]

    if (type === 'video' || type === 'image') {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const { width, height } = stream
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      return { width, height }
    } else if (type === 'audio') {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-argument
      const duration = Math.round(stream.duration || parseWebmDuration(stream))
      return { duration }
    } else {
      throw new Error(`Unknown media type: ${String(type)}`)
    }
  }
