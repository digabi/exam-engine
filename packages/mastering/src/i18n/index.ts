import i18n from 'i18next'
import _ from 'lodash'
import examFootersFi from './exam_footers_fi-FI.json'
import examFootersSv from './exam_footers_sv-FI.json'
import examTitlesFi from './exam_titles_fi-FI.json'
import examTitlesSv from './exam_titles_sv-FI.json'

const resources = {
  'fi-FI': {
    'exam-title': examTitlesFi,
    'exam-footer': examFootersFi,
  },
  'sv-FI': {
    'exam-title': examTitlesSv,
    'exam-footer': examFootersSv,
  },
}

export function initI18n(lng: string): typeof i18n {
  return i18n.createInstance(
    {
      compatibilityJSON: 'v3',
      resources,
      lng,
      fallbackLng: 'fi-FI',
      debug: false,
    },
    _.noop // Force i18next to initialize the instance.
  )
}
