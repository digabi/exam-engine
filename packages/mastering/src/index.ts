export { createMex, createMultiMex } from './createMex'
export * from './mastering/createGradingStructure'
export { getMediaMetadataFromLocalFile } from './getMediaMetadataFromLocalFile'
export * from './mastering'
export { ns } from './mastering/schema'
export * from './migrations'
