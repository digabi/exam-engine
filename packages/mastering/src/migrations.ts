import { Document } from 'libxmljs2'
import { compare } from 'compare-versions'
import { migrateFrom01To02 } from './migrations/migrateFrom01To02'
import { migrateFrom02To03 } from './migrations/migrateFrom02To03'
import { migrateFrom03To04 } from './migrations/migrateFrom03To04'

type Migration = (doc: Document) => void

const noop: Migration = () => {}
const migrations: Record<string, Migration> = {
  '0.1': migrateFrom01To02,
  '0.2': migrateFrom02To03,
  '0.3': migrateFrom03To04,
  '0.4': noop
}
const supportedVersions = Object.keys(migrations)

export const currentExamSchemaVersion = supportedVersions[supportedVersions.length - 1]

/**
 * Tries to migrate an exam to the latest schema version.
 *
 * Note that this function modifies the document.
 */
export function migrateExam(doc: Document): void {
  const examSchemaVersion = doc.root()?.attr('exam-schema-version')?.value()

  if (examSchemaVersion == null) {
    throw new Error(`No exam-schema-version attribute found`)
  } else if (!supportedVersions.includes(examSchemaVersion)) {
    throw new Error(`Unsupported exam schema version: ${examSchemaVersion}.
This version of exam-engine supports exams with the following exam schema versions: ${supportedVersions.join(', ')}`)
  }

  for (const [migrationVersion, migrate] of Object.entries(migrations)) {
    if (compare(migrationVersion, examSchemaVersion, '>=')) {
      migrate(doc)
    }
  }
}
