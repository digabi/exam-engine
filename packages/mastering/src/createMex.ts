import fs from 'fs'
import glob from 'glob-promise'
import path from 'path'
import stream, { Readable, Writable } from 'stream'
import yazl, { ZipFile } from 'yazl'
import { createAES256EncryptStreamWithIv, deriveAES256KeyAndIv, KeyAndIv, signWithSHA256AndRSA } from './crypto-utils'
import cloneable from 'cloneable-readable'
import { promisify } from 'util'

const pipeline = promisify(stream.pipeline)

export interface ExamFile {
  /** A relative filename (e.g. "foo.mp3"). This should be the same filename than in the exam XML. */
  filename: string
  /** A ReadableStream of the file contents */
  contents: Readable
}

export interface AttachmentFile extends ExamFile {
  /** Whether this attachment is restricted or not. */
  restricted: boolean
}

export async function createMex(
  xml: string,
  attachments: AttachmentFile[],
  nsaScripts: Readable,
  securityCodes: Readable | null,
  passphrase: string,
  answersPrivateKey: string,
  outputStream: Writable,
  json?: Buffer | null,
  ktpUpdate?: Readable,
  koeUpdate?: Readable
): Promise<void> {
  const bundleDir = path.dirname(require.resolve('@digabi/exam-engine-core/dist/main-bundle.js'))
  const renderingFiles = await glob(`${bundleDir}/{main-bundle.js,main.css,assets/*}`, {
    nodir: true,
    realpath: true
  })

  const zipFile = new yazl.ZipFile()
  const keyAndIv = deriveAES256KeyAndIv(passphrase)

  encryptAndSign(
    zipFile,
    json ? 'abitti-exam.xml' : 'exam.xml',
    keyAndIv,
    answersPrivateKey,
    toStream(Buffer.from(xml))
  )

  if (json) {
    encryptAndSign(zipFile, 'exam.json', keyAndIv, answersPrivateKey, toStream(json))
  }
  encryptAndSign(zipFile, 'nsa.zip', keyAndIv, answersPrivateKey, nsaScripts)
  if (securityCodes) {
    encryptAndSign(zipFile, 'security-codes.json', keyAndIv, answersPrivateKey, securityCodes)
  }
  if (ktpUpdate) {
    encryptAndSign(zipFile, 'ktp-update.zip', keyAndIv, answersPrivateKey, ktpUpdate)
  }
  if (koeUpdate) {
    const koeUpdateCloneable = cloneable(koeUpdate)
    encryptAndSign(zipFile, 'koe-update.zip', keyAndIv, answersPrivateKey, koeUpdateCloneable.clone())
    sign(zipFile, 'koe-update.zip', answersPrivateKey, koeUpdateCloneable)
  }
  encryptAndSignFiles(
    zipFile,
    'rendering.zip',
    keyAndIv,
    answersPrivateKey,
    renderingFiles.map(renderingFile => ({
      contents: fs.createReadStream(renderingFile),
      filename: path.relative(bundleDir, renderingFile)
    }))
  )
  encryptAndSignFiles(
    zipFile,
    'attachments.zip',
    keyAndIv,
    answersPrivateKey,
    attachments.map(({ filename, contents, restricted }) => ({
      filename: restricted ? path.join('restricted', filename) : filename,
      contents
    }))
  )

  const promise = pipeline(zipFile.outputStream, outputStream)
  zipFile.end()
  await promise
}

export async function createMultiMex(
  exams: ExamFile[],
  nsaScripts: Readable,
  securityCodes: Readable,
  passphrase: string,
  answersPrivateKey: string,
  outputStream: Writable,
  loadSimulationConfiguration?: Readable,
  ktpUpdate?: Readable,
  koeUpdate?: Readable
): Promise<void> {
  const zipFile = new yazl.ZipFile()
  const keyAndIv = deriveAES256KeyAndIv(passphrase)

  for (const exam of exams) {
    zipFile.addReadStream(exam.contents, exam.filename)
  }

  encryptAndSign(zipFile, 'nsa.zip', keyAndIv, answersPrivateKey, nsaScripts)
  encryptAndSign(zipFile, 'security-codes.json', keyAndIv, answersPrivateKey, securityCodes)
  if (loadSimulationConfiguration) {
    encryptAndSign(
      zipFile,
      'load-simulation-configuration.json',
      keyAndIv,
      answersPrivateKey,
      loadSimulationConfiguration
    )
  }
  if (ktpUpdate) {
    encryptAndSign(zipFile, 'ktp-update.zip', keyAndIv, answersPrivateKey, ktpUpdate)
  }
  if (koeUpdate) {
    const koeUpdateCloneable = cloneable(koeUpdate)
    encryptAndSign(zipFile, 'koe-update.zip', keyAndIv, answersPrivateKey, koeUpdateCloneable.clone())
    sign(zipFile, 'koe-update.zip', answersPrivateKey, koeUpdateCloneable)
  }

  const promise = pipeline(zipFile.outputStream, outputStream)
  zipFile.end()
  await promise
}

function encryptAndSignFiles(
  zipFile: ZipFile,
  filename: string,
  keyAndIv: KeyAndIv,
  answersPrivateKey: string,
  files: ExamFile[]
) {
  const innerZipFile = new yazl.ZipFile()
  for (const file of files) {
    innerZipFile.addReadStream(file.contents, file.filename)
  }
  innerZipFile.end()

  encryptAndSign(zipFile, filename, keyAndIv, answersPrivateKey, innerZipFile.outputStream as Readable)
}

function encryptAndSign(
  zipFile: ZipFile,
  filename: string,
  keyAndIv: KeyAndIv,
  answersPrivateKey: string,
  input: Readable
): void {
  const encrypted = cloneable(input.pipe(createAES256EncryptStreamWithIv(keyAndIv)))

  zipFile.addReadStream(encrypted.clone(), `${filename}.bin`)
  sign(zipFile, `${filename}.bin`, answersPrivateKey, encrypted)
}

function sign(zipFile: ZipFile, filename: string, answersPrivateKey: string, input: Readable): void {
  const signer = signWithSHA256AndRSA(input, answersPrivateKey)
  zipFile.addReadStream(signer, `${filename}.sig`)
}

function toStream(buffer: Buffer): Readable {
  const readable = new Readable()
  readable.push(buffer)
  readable.push(null)
  return readable
}
