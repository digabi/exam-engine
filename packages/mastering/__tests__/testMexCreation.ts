import { promises as fs } from 'fs'
import path from 'path'
import yauzl from 'yauzl-promise'
import { Readable, PassThrough } from 'stream'
import { createMex, AttachmentFile, createMultiMex, ExamFile } from '../dist/createMex'
import { verifyWithSHA256AndRSA } from '../src/crypto-utils'

interface InterestingEntryMetadata {
  fileName: string
  uncompressedSize: number
}

describe('Mex exam package creation', () => {
  let privateKey: string
  let publicKey: string
  let nsaScripts: Readable
  let securityCodes: Readable
  let exams: ExamFile[]

  beforeAll(async () => {
    privateKey = await readResource('private-key.pem')
    publicKey = await readResource('public-key.pem')
  })

  beforeEach(() => {
    nsaScripts = Readable.from(['mock nsa scripts'])
    securityCodes = Readable.from(['mock security codes'])
    exams = [
      {
        filename: 'exam_a.mex',
        contents: Readable.from(['mock exam A']),
      },
      {
        filename: 'exam_b.mex',
        contents: Readable.from(['mock exam B']),
      },
    ]
  })

  const xml = 'mock xml'
  const attachments: AttachmentFile[] = []
  const loadSimulationConfiguration = undefined
  const passphrase = 'kellohalli'
  const ktpUpdate = undefined

  function getMexStreamAndBuffers() {
    const mexStream = new PassThrough()
    const mexBuffers: Buffer[] = []
    mexStream.on('data', (data: Buffer) => {
      mexBuffers.push(data)
    })
    return { mexStream, mexBuffers }
  }

  async function expectZipEntriesAreCorrect(
    mexBuffers: Buffer[],
    entryMapper: (value: yauzl.Entry, index: number, array: yauzl.Entry[]) => InterestingEntryMetadata
  ): Promise<yauzl.Entry[]> {
    const zip = await yauzl.fromBuffer(Buffer.concat(mexBuffers))
    const rawEntries = await zip.readEntries()
    const mappedEntries = rawEntries.map(entryMapper)
    expect(mappedEntries).toMatchSnapshot()
    return rawEntries
  }

  describe('with createMex', () => {
    async function expectCorrectMexIsCreated(koeUpdate?: Readable) {
      const { mexStream, mexBuffers } = getMexStreamAndBuffers()
      await createMex(
        xml,
        attachments,
        nsaScripts,
        null,
        passphrase,
        privateKey,
        mexStream,
        undefined,
        ktpUpdate,
        koeUpdate
      )
      return await expectZipEntriesAreCorrect(mexBuffers, (e) => ({
        fileName: e.fileName,
        uncompressedSize: e.fileName === 'rendering.zip.bin' ? 0 : e.uncompressedSize, // ignore rendering.zip size
      }))
    }

    it('creates a minimal mex with verifyable exam.xml signature', async () => {
      const mexEntries = await expectCorrectMexIsCreated()
      const encryptedExamXml = await toBuffer(
        await mexEntries.find((e) => e.fileName === 'exam.xml.bin')!.openReadStream()
      )
      const encryptedExamXmlSignature = (
        await toBuffer(await mexEntries.find((e) => e.fileName === 'exam.xml.bin.sig')!.openReadStream())
      ).toString('utf8')

      const signatureVerificationResult = verifyWithSHA256AndRSA(encryptedExamXml, publicKey, encryptedExamXmlSignature)
      expect(signatureVerificationResult).toBeTruthy()
    })

    it('creates a mex with koe-update.zip', async () => {
      const koeUpdate = Readable.from(['mock koe-update.zip'])
      await expectCorrectMexIsCreated(koeUpdate)
    })
  })

  describe('with createMultiMex', () => {
    async function expectCorrectMultiMexIsCreated(koeUpdate?: Readable) {
      const { mexStream, mexBuffers } = getMexStreamAndBuffers()
      await createMultiMex(
        exams,
        nsaScripts,
        securityCodes,
        passphrase,
        privateKey,
        mexStream,
        loadSimulationConfiguration,
        ktpUpdate,
        koeUpdate
      )
      await expectZipEntriesAreCorrect(mexBuffers, (e) => ({
        fileName: e.fileName,
        uncompressedSize: e.uncompressedSize,
      }))
    }

    it('creates a minimal multimex', async () => {
      await expectCorrectMultiMexIsCreated()
    })

    it('creates a multimex with koe-update.zip', async () => {
      const koeUpdate = Readable.from(['mock koe-update.zip'])
      await expectCorrectMultiMexIsCreated(koeUpdate)
    })
  })
})

async function readResource(filename: string): Promise<string> {
  return fs.readFile(path.resolve(__dirname, 'resources', filename), 'utf-8')
}

function toBuffer(stream: Readable): Promise<Buffer> {
  return new Promise((resolve) => {
    const buffers: Buffer[] = []
    stream.on('data', (d: Buffer) => buffers.push(d))
    stream.on('end', () => resolve(Buffer.concat(buffers)))
  })
}
