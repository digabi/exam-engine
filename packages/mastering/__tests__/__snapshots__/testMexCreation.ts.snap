// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mex exam package creation with createMex creates a mex with koe-update.zip 1`] = `
[
  {
    "fileName": "exam.xml.bin",
    "uncompressedSize": 8,
  },
  {
    "fileName": "exam.xml.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "nsa.zip.bin",
    "uncompressedSize": 16,
  },
  {
    "fileName": "nsa.zip.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "koe-update.zip.bin",
    "uncompressedSize": 19,
  },
  {
    "fileName": "koe-update.zip.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "koe-update.zip.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "rendering.zip.bin",
    "uncompressedSize": 0,
  },
  {
    "fileName": "rendering.zip.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "attachments.zip.bin",
    "uncompressedSize": 22,
  },
  {
    "fileName": "attachments.zip.bin.sig",
    "uncompressedSize": 88,
  },
]
`;

exports[`Mex exam package creation with createMex creates a minimal mex with verifyable exam.xml signature 1`] = `
[
  {
    "fileName": "exam.xml.bin",
    "uncompressedSize": 8,
  },
  {
    "fileName": "exam.xml.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "nsa.zip.bin",
    "uncompressedSize": 16,
  },
  {
    "fileName": "nsa.zip.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "rendering.zip.bin",
    "uncompressedSize": 0,
  },
  {
    "fileName": "rendering.zip.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "attachments.zip.bin",
    "uncompressedSize": 22,
  },
  {
    "fileName": "attachments.zip.bin.sig",
    "uncompressedSize": 88,
  },
]
`;

exports[`Mex exam package creation with createMultiMex creates a minimal multimex 1`] = `
[
  {
    "fileName": "exam_a.mex",
    "uncompressedSize": 11,
  },
  {
    "fileName": "exam_b.mex",
    "uncompressedSize": 11,
  },
  {
    "fileName": "nsa.zip.bin",
    "uncompressedSize": 16,
  },
  {
    "fileName": "nsa.zip.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "security-codes.json.bin",
    "uncompressedSize": 19,
  },
  {
    "fileName": "security-codes.json.bin.sig",
    "uncompressedSize": 88,
  },
]
`;

exports[`Mex exam package creation with createMultiMex creates a multimex with koe-update.zip 1`] = `
[
  {
    "fileName": "exam_a.mex",
    "uncompressedSize": 11,
  },
  {
    "fileName": "exam_b.mex",
    "uncompressedSize": 11,
  },
  {
    "fileName": "nsa.zip.bin",
    "uncompressedSize": 16,
  },
  {
    "fileName": "nsa.zip.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "security-codes.json.bin",
    "uncompressedSize": 19,
  },
  {
    "fileName": "security-codes.json.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "koe-update.zip.bin",
    "uncompressedSize": 19,
  },
  {
    "fileName": "koe-update.zip.bin.sig",
    "uncompressedSize": 88,
  },
  {
    "fileName": "koe-update.zip.sig",
    "uncompressedSize": 88,
  },
]
`;
